### ISSUE TITLE
Feature request, throughput task policy

### ISSUE THREAD

**Issue by bathmatt** [kokkos/kokkos#564]
I have a bunch of tasks which I want to grab the next one until I'm done.  Because of this I would like a throughput policy, therefore, I get all the computing units and start working until my work is done.  On a CPU this is easy, but for a GPU there is no good way to do this.

**Comment by hcedwar** [comment]
Need more info are these tasks the closures of a TeamPolicy?
Are they a heterogeneous collection of closures?

**Comment by bathmatt** [comment]
need more info, what is a closure of a team policy? :)

When I do this I use a bunch of teams.  What I want is, get me a team for every resource and let me use it.  I'll let you know when I'm done.  

In my code I use backfilling to keep a team from diverging, therefore, once a team is done, all the work is done.  So, I want a bunch of teams working until they're all done and a simple way of doing this.  functor::get_max_number_teams() or device::get_max_slots() or something would also work, sort of.

**Comment by hcedwar** [comment]
For the GPU you can just choose the league size to be the number of work items to do, and let the hardware scheduler do the dynamic load balancing work for you.

'closure' a.k.a. functor a.k.a. lambda

**Comment by bathmatt** [comment]
And if you don't know the number of work items you have to do?  But you know there are a lot? Example, things based on random processes.

It would be nice to have all the resources working until one is out of work.

**Comment by hcedwar** [comment]
Ah - dynamic number of work items is the root issue.  For the GPU you can safely set a number larger than the available GPU multiprocessors, say a few hundred.
If you want to use exactly the number of GPU processors:
`Kokkos::Impl::cuda_internal_multiprocessor_count()`
note that this is a `Kokkos::Impl` function that only exists when CUDA is enabled.

**Comment by crtrott** [comment]
This is unfortunately not as easy as it sounds. I though really hard about this. Here are a couple of things:
1) `Kokkos::Impl::cuda_internal_multiprocessor_count()` is not enough because you need to take into account how many teams can be scheduled per multi processor. 
2) If another process shares the same GPU this might get tricky if you truly need concurrently scheduled teams. I.e. we can not guarantee that teams will run concurrently, even if the number of teams would theoretically fit on the GPU. 
3) Similar issue as 2): if we have multiple control threads sharing the same GPU the same issue arises. 

Matt: your problem seems somewhat simpler. If you actually do have an internal workqueue you could just lunch more than you need. But no team will return before all the work is done. In the end you will have quite a few teams starting up to find out everybody else is done and exit again, but this way you get maximal utilization.

**Comment by bathmatt** [comment]
So, how much time do I lose with this idea?  What is the cost of a null
functor on GPU?

I don't have any idea on a number of teams to ask for?  100, 1000, 1M?

**Comment by hcedwar** [comment]
Go for 1000, or 10 * Kokkos::Impl::cuda_internal_multiprocessor_count()

**Comment by crtrott** [comment]
The time you loose should be minimally. And you can query ExecSpace::concurrency()/team_size, and if you use Kokkos::AUTO use 128 for team_size.

On Kepler you can compare that to running 120 teams.
